### DISCUSSION TITLE
delay between tests that use DropwizardExtensionsSupport

### DISCUSSION THREAD

**Discussion by Athou** [dropwizard/dropwizard#8176] [General]
Hey there,

I've been using dropwizard for [CommaFeed](https://github.com/Athou/commafeed/) for almost 10 years now, thank you for the great work, it's been a joy to work with! 😃 

I updated to 4.0.5 from 2.1.6 and everything works fine. However, I just noticed that the integration tests take a lot more time now, there's a ~2s delay between tests during server shutdown that was not happening before:

```
INFO  [2024-01-05 21:25:35,322] org.eclipse.jetty.server.Server: Stopped Server@472dc215{STOPPING}[11.0.19,sto=30000]
INFO  [2024-01-05 21:25:35,322] org.eclipse.jetty.server.Server: Shutdown Server@472dc215{STOPPING}[11.0.19,sto=30000]
INFO  [2024-01-05 21:25:37,327] org.eclipse.jetty.server.AbstractConnector: Stopped application@696a6506{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
INFO  [2024-01-05 21:25:37,332] org.eclipse.jetty.server.handler.ContextHandler: Stopped i.d.j.MutableServletContextHandler@6ed238c2{/,null,STOPPED}
```

The regression is noticeable since this commit https://github.com/Athou/commafeed/commit/fd0e5426e57c42f83860c5deb338a41165699ef3
The delay does not seem to happen when the app is shutdown when running normally, it only seem to happen when tests annotated with `@ExtendWith(DropwizardExtensionsSupport.class)` are running.

Is this something you noticed too? Do you have any pointers as to where to look or why this is happening?

Thanks!

**Comment by Athou** [comment]
Alright so I slept on it and enabled debugging:

```
INFO  [2024-01-06 07:26:44,205] org.eclipse.jetty.server.Server: Shutdown Server@a15e3c1{STOPPING}[11.0.19,sto=30000]
DEBUG [2024-01-06 07:26:44,206] org.eclipse.jetty.server.Server: graceful application@16448827{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
DEBUG [2024-01-06 07:26:44,206] org.eclipse.jetty.server.Server: graceful i.d.j.MutableServletContextHandler@6f1c3f18{/,null,AVAILABLE}
DEBUG [2024-01-06 07:26:44,206] org.eclipse.jetty.server.Server: graceful StatisticsHandler@608b74f3{STARTED,r=0,d=0}
DEBUG [2024-01-06 07:26:44,206] org.eclipse.jetty.server.Server: graceful SessionTracker@7ec49d9b{STARTED}
DEBUG [2024-01-06 07:26:44,207] org.eclipse.jetty.util.thread.QueuedThreadPool: ran pool-10-thread-12-acceptor-1@2ef4e7f3-application@16448827{HTTP/1.1, (http/1.1)}{0.0.0.0:8080} in InstrumentedQueuedThreadPool[dw]@1ebcfcf1{STARTED,8<=14<=1024,i=1,r=-1,t=59527ms,q=0}[ReservedThreadExecutor@100d5018{reserved=1/20,pending=0}]
DEBUG [2024-01-06 07:26:44,207] org.eclipse.jetty.util.thread.QueuedThreadPool: ran pool-10-thread-11-acceptor-0@3bf47d1e-application@16448827{HTTP/1.1, (http/1.1)}{0.0.0.0:8080} in InstrumentedQueuedThreadPool[dw]@1ebcfcf1{STARTED,8<=14<=1024,i=1,r=-1,t=59527ms,q=0}[ReservedThreadExecutor@100d5018{reserved=1/20,pending=0}]
DEBUG [2024-01-06 07:26:44,207] org.eclipse.jetty.util.thread.QueuedThreadPool: Evict check, period=60000ms InstrumentedQueuedThreadPool[dw]@1ebcfcf1{STARTED,8<=14<=1024,i=3,r=-1,t=59527ms,q=0}[ReservedThreadExecutor@100d5018{reserved=1/20,pending=0}]
DEBUG [2024-01-06 07:26:44,207] org.eclipse.jetty.util.thread.QueuedThreadPool: Evict check, period=60000ms InstrumentedQueuedThreadPool[dw]@1ebcfcf1{STARTED,8<=14<=1024,i=3,r=-1,t=59527ms,q=0}[ReservedThreadExecutor@100d5018{reserved=1/20,pending=0}]
DEBUG [2024-01-06 07:26:44,207] org.eclipse.jetty.util.thread.QueuedThreadPool: Evict skipped, threshold=119527ms in the future InstrumentedQueuedThreadPool[dw]@1ebcfcf1{STARTED,8<=14<=1024,i=3,r=-1,t=59527ms,q=0}[ReservedThreadExecutor@100d5018{reserved=1/20,pending=0}]
DEBUG [2024-01-06 07:26:44,207] org.eclipse.jetty.util.thread.QueuedThreadPool: Evict skipped, threshold=119527ms in the future InstrumentedQueuedThreadPool[dw]@1ebcfcf1{STARTED,8<=14<=1024,i=3,r=-1,t=59527ms,q=0}[ReservedThreadExecutor@100d5018{reserved=1/20,pending=0}]
DEBUG [2024-01-06 07:26:44,207] org.eclipse.jetty.io.IdleTimeout: Setting idle timeout 30000 -> 1000 on SocketChannelEndPoint@60b32495[{l=/127.0.0.1:8080,r=/127.0.0.1:58811,OPEN,fill=FI,flush=-,to=4/1000}{io=1/1,kio=1,kro=1}]->[HttpConnection@528ca39c[p=HttpParser{s=START,0 of -1},g=HttpGenerator@7bb75668{s=START}]=>HttpChannelOverHttp@6e338fa8{s=HttpChannelState@9adf938{s=IDLE rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=2,c=false/false,a=IDLE,uri=null,age=0}]
DEBUG [2024-01-06 07:26:44,207] org.eclipse.jetty.io.IdleTimeout: SocketChannelEndPoint@60b32495[{l=/127.0.0.1:8080,r=/127.0.0.1:58811,OPEN,fill=FI,flush=-,to=4/1000}{io=1/1,kio=1,kro=1}]->[HttpConnection@528ca39c[p=HttpParser{s=START,0 of -1},g=HttpGenerator@7bb75668{s=START}]=>HttpChannelOverHttp@6e338fa8{s=HttpChannelState@9adf938{s=IDLE rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=2,c=false/false,a=IDLE,uri=null,age=0}] idle timeout check, elapsed: 4 ms, remaining: 996 ms
DEBUG [2024-01-06 07:26:45,204] org.eclipse.jetty.io.IdleTimeout: SocketChannelEndPoint@60b32495[{l=/127.0.0.1:8080,r=/127.0.0.1:58811,OPEN,fill=FI,flush=-,to=1001/1000}{io=1/1,kio=1,kro=1}]->[HttpConnection@528ca39c[p=HttpParser{s=START,0 of -1},g=HttpGenerator@7bb75668{s=START}]=>HttpChannelOverHttp@6e338fa8{s=HttpChannelState@9adf938{s=IDLE rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=2,c=false/false,a=IDLE,uri=null,age=0}] idle timeout check, elapsed: 1001 ms, remaining: -1 ms
DEBUG [2024-01-06 07:26:45,204] org.eclipse.jetty.io.IdleTimeout: SocketChannelEndPoint@60b32495[{l=/127.0.0.1:8080,r=/127.0.0.1:58811,OPEN,fill=FI,flush=-,to=1001/1000}{io=1/1,kio=1,kro=1}]->[HttpConnection@528ca39c[p=HttpParser{s=START,0 of -1},g=HttpGenerator@7bb75668{s=START}]=>HttpChannelOverHttp@6e338fa8{s=HttpChannelState@9adf938{s=IDLE rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=2,c=false/false,a=IDLE,uri=null,age=0}] idle timeout expired
DEBUG [2024-01-06 07:26:45,204] org.eclipse.jetty.io.FillInterest: onFail FillInterest@7e88356{ReadCallback@242e49e4{HttpConnection@528ca39c::SocketChannelEndPoint@60b32495[{l=/127.0.0.1:8080,r=/127.0.0.1:58811,OPEN,fill=FI,flush=-,to=1002/1000}{io=1/1,kio=1,kro=1}]->[HttpConnection@528ca39c[p=HttpParser{s=START,0 of -1},g=HttpGenerator@7bb75668{s=START}]=>HttpChannelOverHttp@6e338fa8{s=HttpChannelState@9adf938{s=IDLE rs=BLOCKING os=OPEN is=IDLE awp=false se=false i=true al=0},r=2,c=false/false,a=IDLE,uri=null,age=0}]}}
! java.util.concurrent.TimeoutException: Idle timeout expired: 1001/1000 ms
! at org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:170)
! at org.eclipse.jetty.io.IdleTimeout.idleCheck(IdleTimeout.java:112)
! at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
! at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
! at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
! at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
! at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
! at java.base/java.lang.Thread.run(Thread.java:833)
```

It seems that jetty is waiting for HTTP connections to close before shutdown. It happens twice in the log (1s for each connection so it explains my 2s delay).
I wrote a test that doesn't use the extension HTTP client and it shuts down immediately, so there seems to be dangling idle connections.

[I added a cleanup method](https://github.com/Athou/commafeed/commit/60b6c69020ed67abe81621463a494a7d1816198a) to close the HTTP client after each test and that solves the issue, the tests now run as fast as before.

```java
@AfterEach
void cleanup() {
    this.client.close();
}
```

Am I doing something wrong in my tests? I don't think I'm leaving open connections.

Should the documentation be updated to include a cleanup method, or should `DropwizardExtensionsSupport` maybe be modified to close the HTTP client after each test?

**Comment by zUniQueX** [comment]
Hi @Athou. The problem you describe is related to the handling of HTTP connections. As the [Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection?retiredLocale=de#any_comma-separated_list_of_http_headers_usually) state, the default value for the `Connection` header for HTTP/1.1 is `keep-alive`. So the connections won't be closed immediately. If you call `client.close()`, the client will terminate the open connections, so Jetty won't wait for expiration.

Regarding the exception in the Jetty debug logs: The Jetty `idleTimeout` property controls the keep-alive duration. This duration can be set in the [connector properties](https://www.dropwizard.io/en/stable/manual/configuration.html#http). If you set the value to `1ms`, the server will shut down immediately without the client having to be closed.

However, the much cleaner option for Dropwizard is to set this with a client property. Dropwizard 3.x and 4.x use the `Apache5ConnectorProvider`. For that provider there is the property `Apache5ClientProperties.REUSE_STRATEGY` that can be set to control the keep-alive behavior. When you add the following line to your test setup, the application should shut down immediately after finishing the test:

```java
client.property(Apache5ClientProperties.REUSE_STRATEGY, (ConnectionReuseStrategy) (request, response, context) -> false);
```

That being said, the behavior to keep connections alive is common practice and shouldn't be changes for the testing package IMHO. With this viable workaround, that should be fine.

**Comment by easbar** [comment]
We ran into a similar problem when trying to upgrade https://github.com/graphhopper/graphhopper/pull/2920 from Dropwizard 2 to Dropwizard 3. Some of our tests were hanging indefinitely after the upgrade. We were able to reduce the problem to the following code (run within the Dropwizard repository on the release/4.0.x branch):

```java
package io.dropwizard.testing.junit5;

import io.dropwizard.core.Application;
import io.dropwizard.core.Configuration;
import io.dropwizard.core.setup.Environment;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.client.Entity;
import jakarta.ws.rs.core.Response;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith(DropwizardExtensionsSupport.class)
public class DropwizardAppExtensionClientHangsTest {

    @Path("/")
    public static class MyResource {
        @POST
        public Response doPost() {
            return Response.ok().build();
        }
    }

    public static class MyApp extends Application<Configuration> {

        @Override
        public void run(Configuration configuration, Environment environment) throws Exception {
            environment.jersey().register(new MyResource());
        }
    }

    private static final DropwizardAppExtension<Configuration> app = new DropwizardAppExtension<>(MyApp.class, new Configuration());

    @Test
    public void myTest() {
        for (int i = 0; i < 10; i++) {
            post();
            System.out.println("POST " + i + " complete");
        }
    }

    private static void post() {
        // send POST request against a missing endpoint (provoking 404)
        Response response = app.client().target("http://localhost:8080/not_found").request().post(Entity.json("{}"));
        System.out.println("status: " + response.getStatus());
    }
}
```

This test hangs after completing the first ~6 requests. The same happens on the release/3.0.x branch, but the release/2.0.x branch does not show this problem.

Adding

```java
        app.client().property(Apache5ClientProperties.REUSE_STRATEGY, (ConnectionReuseStrategy) (request, response, context) -> false);
``` 

before sending the request in the `post()` method seems to fix the issue, but it seems a bit odd that this is now necessary?

**Reply by joschi** [reply]
Dropwizard 3.x/4.x migrated `dropwizard-client` to Apache HttpClient 5.x, so this could just be a "regular" regression: https://www.dropwizard.io/en/release-3.0.x/manual/upgrade-notes/upgrade-notes-3_0_x.html#apache-httpclient-5

Anyway, I think it might make sense to set this configuration property by default in `dropwizard-testing`.

@zUniQueX Any thoughts on that?

**Reply by otbutz** [reply]
Maybe version 5.x isn't so lenient anymore? I can't find the 5.X equivalent, but the 4.5.X documentation is pretty straightforward with the requirements

> In order to ensure proper release of system resources one **must** close either the content stream associated with the entity or the response itself

> The difference between closing the content stream and closing the response is that the former will attempt to keep the underlying connection alive by consuming the entity content while the latter immediately shuts down and discards the connection.

https://hc.apache.org/httpcomponents-client-4.5.x/current/tutorial/html/fundamentals.html#d5e145

So either

```java
if (response.getEntity() != null) {
    response.getEntity().getContent().close();
}
```

or to close the entire connection:

```java
response.close();
```

**Reply by zUniQueX** [reply]
@joschi I don't think it's a good option to set the client property to fix this.

I can't find an issue on the client-side. We don't use `dropwizard-client` in the `dropwizard-testing` module to avoid circular dependencies. However, we've made an upgrade from the `GrizzlyConnector` to the `Apache5Connector` while preparing the upgrade for Dropwizard 3.x.

When using the `Apache5Connector` with a Dropwizard 2.x application, the delay **is not** reproducible. When using the `GrizzlyConnector` with a Dropwizard 3.x application, the issue **is** reproducible. So this is more likely a server-side issue, when both clients behave equally on the different Dropwizard versions. Setting a client property to fix this seems odd for me.

**Reply by zUniQueX** [reply]
@otbutz When using the Jersey `Apache5Connector` the stream and the response are closed by default: https://github.com/eclipse-ee4j/jersey/blob/2.x/connectors/apache5-connector/src/main/java/org/glassfish/jersey/apache5/connector/Apache5ConnectionClosingStrategy.java#L77-L83

**Reply by karussell** [reply]
@joschi do you have new information regarding this issue?

I used the snippet from @easbar (with sometimes slightly different imports) and ran it against different releases using the command: `mvn clean install -pl dropwizard-testing`. It seems that it is fine for `v3.0.0-beta.3` but hangs for `v3.0.0-beta.4` (see [the changes](https://github.com/dropwizard/dropwizard/compare/v3.0.0-beta.3...v3.0.0-beta.4)).

The argumentation from @zUniQueX makes sense to me and so I thought this sounds like a server-side problem and would be an argument not to migrate yet.

But from the linked changes I read the section "Modification of the client in ``DropwizardAppExtension``" and so I replaced `app.client()` with `client = new JerseyClientBuilder().build()` and then it did no longer hang with `v3.0.0-beta.4`. So this sounds more like a client issue!?

**Reply by karussell** [reply]
@joschi sorry to ping you again on this issue. It is tricky for us to decide from the outside if this is a client or server issue. Could you have a quick look or was this already fixed in a more recent version?

**Reply by karussell** [reply]
I tried v3.0.0-beta.3 but used [the dependencies](https://github.com/dropwizard/dropwizard/blob/c9c8d72708fbfc8b9982e9b37f321e588a1efdab/dropwizard-dependencies/pom.xml) of [v3.0.0-beta.4](https://github.com/dropwizard/dropwizard/blob/f679616d7ad7df6ba3bb77535507ca8ec4049336/dropwizard-dependencies/pom.xml) and the test does not hang.

There is also no problem when I switch to [this commit](https://github.com/dropwizard/dropwizard/commit/c956596ef5da8478de1837d13da1d3719b8e00ed) right before the httpClient change. 

But the test hangs with [this commit that switches to httpClient 5](https://github.com/dropwizard/dropwizard/commit/fb48a5c8af9e5b49e6fa83301116711454cd4671).

So, this together points more to a client issue IMO or what do other think of this?

Let's assume this is a client issue shouldn't it be avoided with e.g. a client config change in DropwizardAppExtension.java?
